### ISSUE TITLE
AuditTrailSettingsUpdateTests#testDynamicFilterSettings fails due to unknown role

### ISSUE THREAD

**Issue by benwtrent** [elastic/elasticsearch#69693]
**Build scan**:
https://gradle-enterprise.elastic.co/s/rctzxpq6cf5va

**Repro line**:

`./gradlew ':x-pack:plugin:security:internalClusterTest' --tests "org.elasticsearch.xpack.security.audit.logfile.AuditTrailSettingsUpdateTests.testDynamicFilterSettings" -Dtests.seed=E6DF54FCB75BE6A4 -Dtests.security.manager=true -Dtests.locale=zh-SG -Dtests.timezone=Atlantic/Stanley -Druntime.java=8`

**Reproduces locally?**:
no
**Applicable branches**:
7.x
**Failure history**:
one previous failure so far: https://gradle-enterprise.elastic.co/s/acujcu2t6wbbm
**Failure excerpt**:

```

org.elasticsearch.xpack.security.audit.logfile.AuditTrailSettingsUpdateTests > testDynamicFilterSettings FAILED |  
-- | --
  | java.lang.IllegalArgumentException: unknown role [data_warm] |  
  | at org.elasticsearch.cluster.node.DiscoveryNode.getRoleFromRoleName(DiscoveryNode.java:509) |  
  | at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) |  
  | at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384) |  
  | at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) |  
  | at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) |  
  | at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) |  
  | at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) |  
  | at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) |  
  | at org.elasticsearch.common.settings.Setting.lambda$listSetting$45(Setting.java:1517) |  
  | at org.elasticsearch.common.settings.Setting.get(Setting.java:457) |  
  | at org.elasticsearch.common.settings.Setting.get(Setting.java:451) |  
  | at org.elasticsearch.cluster.node.DiscoveryNode.getRolesFromSettings(DiscoveryNode.java:223) |  
  | at org.elasticsearch.cluster.node.DiscoveryNode.isDataNode(DiscoveryNode.java:82) |  
  | at org.elasticsearch.test.InternalTestCluster.lambda$static$0(InternalTestCluster.java:192) |  
  | at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174) |  
  | at java.util.TreeMap$ValueSpliterator.forEachRemaining(TreeMap.java:2897) |  
  | at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) |  
  | at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) |  
  | at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) |  
  | at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) |  
  | at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) |  
  | at org.elasticsearch.test.InternalTestCluster.filterNodes(InternalTestCluster.java:2374) |  
  | at org.elasticsearch.test.InternalTestCluster.dataNodeAndClients(InternalTestCluster.java:2365) |  
  | at org.elasticsearch.test.InternalTestCluster.numDataNodes(InternalTestCluster.java:2313) |  
  | at org.elasticsearch.test.ESIntegTestCase.maximumNumberOfReplicas(ESIntegTestCase.java:687) |  
  | at org.elasticsearch.test.ESIntegTestCase.numberOfReplicas(ESIntegTestCase.java:692) |  
  | at org.elasticsearch.test.ESIntegTestCase.randomIndexTemplate(ESIntegTestCase.java:413) |  
  | at org.elasticsearch.test.ESIntegTestCase.lambda$beforeInternal$0(ESIntegTestCase.java:372) |  
  | at org.elasticsearch.test.ESIntegTestCase.beforeInternal(ESIntegTestCase.java:383) |  
  | at org.elasticsearch.test.ESIntegTestCase.setupTestCluster(ESIntegTestCase.java:2236) |  
  | at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) |  
  | at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) |  
  | at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) |  
  | at java.lang.reflect.Method.invoke(Method.java:498) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:972) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988) |  
  | at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49) |  
  | at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45) |  
  | at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48) |  
  | at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64) |  
  | at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368) |  
  | at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817) |  
  | at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894) |  
  | at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48) |  
  | at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41) |  
  | at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40) |  
  | at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53) |  
  | at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47) |  
  | at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64) |  
  | at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368) |  
  | at java.lang.Thread.run(Thread.java:748) |  
  |   |  
  | java.lang.IllegalArgumentException: unknown role [data_warm] |  
  | at org.elasticsearch.cluster.node.DiscoveryNode.getRoleFromRoleName(DiscoveryNode.java:509) |  
  | at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) |  
  | at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384) |  
  | at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) |  
  | at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) |  
  | at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) |  
  | at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) |  
  | at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) |  
  | at org.elasticsearch.common.settings.Setting.lambda$listSetting$45(Setting.java:1517) |  
  | at org.elasticsearch.common.settings.Setting.get(Setting.java:457) |  
  | at org.elasticsearch.common.settings.Setting.get(Setting.java:451) |  
  | at org.elasticsearch.cluster.node.DiscoveryNode.getRolesFromSettings(DiscoveryNode.java:223) |  
  | at org.elasticsearch.cluster.node.DiscoveryNode.isDataNode(DiscoveryNode.java:82) |  
  | at org.elasticsearch.indices.cluster.IndicesClusterStateService.doStop(IndicesClusterStateService.java:188) |  
  | at org.elasticsearch.common.component.AbstractLifecycleComponent.stop(AbstractLifecycleComponent.java:68) |  
  | at org.elasticsearch.node.Node.stop(Node.java:940) |  
  | at org.elasticsearch.node.Node.close(Node.java:970) |  
  | at org.elasticsearch.test.InternalTestCluster$NodeAndClient.close(InternalTestCluster.java:1077) |  
  | at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:74) |  
  | at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:116) |  
  | at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:99) |  
  | at org.elasticsearch.test.InternalTestCluster.close(InternalTestCluster.java:869) |  
  | at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:593) |  
  | at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:2253) |  
  | at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) |  
  | at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) |  
  | at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) |  
  | at java.lang.reflect.Method.invoke(Method.java:498) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996) |  
  | at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49) |  
  | at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45) |  
  | at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48) |  
  | at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64) |  
  | at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368) |  
  | at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817) |  
  | at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883) |  
  | at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894) |  
  | at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48) |  
  | at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41) |  
  | at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40) |  
  | at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53) |  
  | at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47) |  
  | at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64) |  
  | at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368) |  
  | at java.lang.Thread.run(Thread.java:748)

```

I wonder if this is related to: https://github.com/elastic/elasticsearch/pull/68226

**Comment by romseygeek** [comment]
I was just pinged with a similar failure in a different test:
https://gradle-enterprise.elastic.co/s/gp25pqqx5zumq/console-log?task=:x-pack:plugin:security:internalClusterTest

**Comment by droberts195** [comment]
> I wonder if this is related to: #68226

If it is then that would mean:
1. In the tests that are failing the transport client is being used to communicate with the internal test clusters
2. Since the change was to add knowledge of extra roles to the transport client, the problem must now be that the client knows about more roles than the server in the failure cases

Was randomization added somewhere so that internal cluster tests sometimes give data nodes the `data` role and other times the `data_warm` role?  But without loading the X-Pack core plugin into the internal test cluster?  Given the changes of #68226 that is the only way I can think of that it could be related.

**Comment by benwtrent** [comment]
https://gradle-enterprise.elastic.co/s/lfmokptlo5wdc

Is another similar failure: 

```

org.elasticsearch.repositories.encrypted.EncryptedS3BlobStoreRepositoryIntegTests > testSnapshotAndRestore FAILED |  
-- | --
  | java.lang.IllegalArgumentException: unknown role [data_frozen] |  
  | at org.elasticsearch.cluster.node.DiscoveryNode.getRoleFromRoleName(DiscoveryNode.java:509) |  
  | at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) |  
  | at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655) |  
  | at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) |  
  | at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) |  
  | at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) |  
  | at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) |  
  | at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) |  
  | at org.elasticsearch.common.settings.Setting.lambda$listSetting$45(Setting.java:1517)


```

**Comment by jakelandis** [comment]
not sure if this is a security area, or a core features issue. Unknown role "data_frozen" <-- @dakrone can you triage this ?

**Comment by droberts195** [comment]
Given the frequency this is now occurring at it must have been caused by #68226.  So the easy way out would be to revert that.  However, it would be a shame to do that without understanding why adding extra role knowledge to the transport client makes these internal cluster tests fail, because it definitely fixed a problem that users of the transport client might be having when connecting to real clusters.

**Comment by droberts195** [comment]
I reverted the change of #68226 in 914885b.  Hopefully that will stop the failures.

**Comment by rjernst** [comment]
@droberts195 Could it be this (in InternalTestCluster)?

```
/*
 * The node.roles setting does not make sense for the transport client, filter it. If the transport client were not deprecated
 * we would probably want to invest in infrastructure to mark a setting as not applicable to the transport client and then
 * filter all such settings here.
 */
final Settings finalSettings = builder.build().filter(k -> k.equals(NodeRoleSettings.NODE_ROLES_SETTING.getKey()) == false);
```

**Comment by droberts195** [comment]
Thanks @rjernst that was a good clue to see what's happening.

It turns out the changes I added in #68226 were already in the first attempt at adding pluggable roles, but then were reverted in #64276 due to #64260.  So it seems that a decision had already been made to leave the transport client unable to work with certain pluggable node roles due to the problem of static variables making it impossible to get the internal cluster tests to work with the pluggable roles.  So by reverting #68226 we are back to that state.

The failures should stop once the revert has passed through intake and now the problem is understood I'll close this.
